charmbits/httpservice/service.go

```diff
@@ -208,191 +327,120 @@ func (svc *Service) Restart() error {
-	listener, err := net.Listen("tcp", addr)
-	if err != nil {
-		return nil, "", errgo.Newf("cannot listen on %s: %v", addr, err)
-	}
-	return tls.NewListener(
-		tcpKeepAliveListener{listener.(*net.TCPListener)},
-		config,
-	), addr, nil
-}
-
-type handler struct {
-	argType reflect.Type
-	fv      reflect.Value
+type handlerInfo struct {
+	argType      reflect.Type
+	relationType reflect.Type
+	fv           reflect.Value
 }
 
 var (
-	httpHandlerType = reflect.TypeOf((*http.Handler)(nil)).Elem()
-	errorType       = reflect.TypeOf((*error)(nil)).Elem()
+	handlerType = reflect.TypeOf((*Handler)(nil)).Elem()
+	errorType   = reflect.TypeOf((*error)(nil)).Elem()
 )
 
-func newHandler(f interface{}) (*handler, error) {
+// newHandler makes a new handler value from the given function,
+// which should be in one of the forms expected by Service.Register.
+// It also registers any relations required by the handler.
+func (svc *Service) newHandlerInfo(f interface{}, registry *hook.Registry) (*handlerInfo, error) {
 	fv := reflect.ValueOf(f)
 	ft := fv.Type()
 	if ft.Kind() != reflect.Func {
-		return nil, errgo.Newf("bad handler; got %T, expected function", f)
+		return nil, errgo.Newf("bad handler: got %T, expected function", f)
 	}
-	if n := ft.NumIn(); n != 1 {
-		return nil, errgo.Newf("bad handler; got %d arguments, expected 1", n)
+	if n := ft.NumIn(); n != 1 && n != 2 {
+		return nil, errgo.Newf("bad handler:ag got %d arguments, expected 1 or 2", n)
 	}
 	if n := ft.NumOut(); n != 2 {
-		return nil, errgo.Newf("bad handler; got %d return values, expected 2", n)
+		return nil, errgo.Newf("bad handler: got %d return values, expected 2", n)
 	}
-	if ft.Out(0) != httpHandlerType || ft.Out(1) != errorType {
-		return nil, errgo.Newf("bad handler; got return values (%s, %s), expected (http.Handler, error)", ft.Out(0), ft.Out(1))
+	if ft.Out(0) != handlerType || ft.Out(1) != errorType {
+		return nil, errgo.Newf("bad handler: got return values (%s, %s), expected (httpservice.Handler, error)", ft.Out(0), ft.Out(1))
 	}
-	return &handler{
+	h := &handlerInfo{
 		argType: ft.In(0),
 		fv:      fv,
-	}, nil
+	}
+	if ft.NumIn() == 1 {
+		// No relations, nothing more to do.
+		return h, nil
+	}
+	svc.relationValues = make(map[string][]byte)
+	rt := ft.In(1)
+	if rt.Kind() != reflect.Ptr || rt.Elem().Kind() != reflect.Struct {
+		return nil, errgo.Newf("bad handler: second argument is %v not a pointer to struct", h.relationType)
+	}
+	h.relationType = rt.Elem()
+	getters := make(map[string]func() (interface{}, error))
+	for i := 0; i < h.relationType.NumField(); i++ {
+		// TODO anonymous fields?
+		f := h.relationType.Field(i)
+		if f.PkgPath != "" {
+			continue
+		}
+		reg := registeredRelations[f.Type]
+		if reg == nil {
+			return nil, errgo.Newf("bad handler: field %s of type %s is not a registered relation type: missing import?", f.Name, f.Type)
+		}
+		getters[f.Name] = reg.registerField(registry.Clone("rel-"+f.Name), f.Tag.Get("httpservice"))
```

Comment:
The only thing that matters with the Clone arguments is that every sub-registry must have a unique name. The actual name itself doesn't matter, but can be seen in the names of the state directories on disk, so the "rel-" is just for that.